[cutlass backend] Change serialization protocol to use more json and cache

henrylhtsang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #157866
* __->__ #157840
* #157839

Differential Revision: [D77949177](https://our.internmc.facebook.com/intern/diff/D77949177/)

What this diff does:
* use lru_cache for serialization and deserialization 
* json dumps more. This seems to help perf.

For instantiation level 3332, the loading time decreases from 33s to 20s (roughly 40%) decrease.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov

henrylhtsang committed:
[cutlass backend] Change serialization protocol to use more json and cache



Differential Revision: [D77949177](https://our.internmc.facebook.com/intern/diff/D77949177/)

[ghstack-poisoned]

henrylhtsang committed:
Update on "[cutlass backend] Change serialization protocol to use more json and cache"


Differential Revision: [D77949177](https://our.internmc.facebook.com/intern/diff/D77949177/)

What this diff does:
* use lru_cache for serialization and deserialization 
* json dumps more. This seems to help perf.

For instantiation level 3332, the loading time decreases from 33s to 20s (roughly 40%) decrease.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

henrylhtsang committed:
Update on "[cutlass backend] Change serialization protocol to use more json and cache"


Differential Revision: [D77949177](https://our.internmc.facebook.com/intern/diff/D77949177/)

What this diff does:
* use lru_cache for serialization and deserialization 
* json dumps more. This seems to help perf.

For instantiation level 3332, the loading time decreases from 33s to 20s (roughly 40%) decrease.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

ColinPeppler commented on test/inductor/test_cutlass_backend.py:
```diff
@@ -2018,6 +2005,37 @@ def linear(x_fp8, x_inverse_scale, w_t_fp8, w_inverse_scale, bias):
+                args, _ = sa.call_args
+                _, choices, _, __ = args
+
+                self.assertEqual(len(choices), 78)
```
would be better to set 78 to a constant variable. 78 looks arbitrary

ColinPeppler commented on torch/_inductor/codegen/cuda/serialization.py:
```diff
@@ -295,172 +270,238 @@ def _json_to_tile_description(cls, json_dict):
     if not try_import_cutlass():
         return None
-    return CUTLASSOperationSerializer()
+    return CUTLASSOperationSerializer
```
todo: fix linter

henrylhtsang committed:
Update on "[cutlass backend] Change serialization protocol to use more json and cache"


Differential Revision: [D77949177](https://our.internmc.facebook.com/intern/diff/D77949177/)

What this diff does:
* use lru_cache for serialization and deserialization 
* json dumps more. This seems to help perf.

For instantiation level 3332, the loading time decreases from 33s to 20s (roughly 40%) decrease.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]
